modules: lvgl: add keypad input device

faxe1008 opened the pull request:
PR contains implementation of the last remaining LVGL input device, the keypad. The keypad can be used to navigate between widgets. Also if a textarea is focused the keypad can be used essentially as a keyboard for inputing characters. 

The keyboard like functionality can be enabled via Kconfig. For conversion between the input-event-codes I took some inspiration from `kbd loadkeys`. They can be generated with `loadkeys us-acentos --mktable`. 

**Things that still need to be added:**
- A sample which used the new keypad input device. I'd love @kartben idea to have a GUI shell interface.

**Things that I'd love some feedback on:**

- Whats the best way of getting the LVGL codes into the device tree, instead of writing `0x10`. Including `<lvgl.h>` is possible but is it feasable?

faxe1008 committed:
modules: lvgl: input: add zephyr,lvgl-keypad-input device binding

Add a pseudo device which can be used to hook into input events and
emit lvgl keypad events.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

faxe1008 commented on modules/lvgl/Kconfig.input:
```diff
@@ -77,4 +77,45 @@ config LV_Z_ENCODER_INPUT_MSGQ_COUNT
+endchoice
+
+config LV_Z_KEYPAD_KEYMAP_SHIFT_MODIFIER
+		bool "Add keymap for shift modifier key."
```
Indent

faxe1008 commented on modules/lvgl/input/keymap/keymap_de_qwertz.c:
```diff
@@ -0,0 +1,54 @@
+#endif /* CONFIG_LV_Z_KEYPAD_KEYMAP_SHIFT_MODIFIER */
+
+uint8_t *key_maps[] = {plain_map,
+		       COND_CODE_1(CONFIG_LV_Z_KEYPAD_KEYMAP_SHIFT_MODIFIER, (shift_map), (0))};
```
newline missing

fabiobaltieri commented:
> * In theory LVGL supports a `uint32_t` keycode meaning that unicode is supported. However this would increase the memory consumption of the keymaps.

Well, not much you can do really, either you figure a way to calculate the code or burn flash on a map, the current one are very sparse though, don't think it's a good idea.

> * Whats the best way of getting the LVGL codes into the device tree, instead of writing `0x10`. Including `<lvgl.h>` is possible but is it feasable?

Not really, you can't have any struct or enum there, just plain defines, so you'd have to rewrite a new file just for that.

> * The naming schema of the keymaps.

How about pulling back on the keymaps for now and start with an input event -> key mapping?

I'm wrapping some rework of the keyboard scan drivers, these right now only generate row-col-touch sets though, so the next step would be to figure out a generic matrix to event mapping strategy anyway, I think we can defer to that and try to see a common schema that we can use for both HID and LVGL. ZMK has something in place we can use as a starting point https://github.com/zmkfirmware/zmk/blob/main/app/boards/arm/nice60/nice60.keymap

faxe1008 commented:
Heyhey @fabiobaltieri 

> Not really, you can't have any struct or enum there, just plain defines, so you'd have to rewrite a new file just for that.

Will do that then, seems to be worth the extra header imo.

> How about pulling back on the keymaps for now and start with an input event -> key mapping?

>  I think we can defer to that and try to see a common schema that we can use for both HID and LVGL. ZMK has something in place we can use as a starting point

👍 on that. I was just drafting a possible solution to see where it goes. Having a general solution for both the kscan and LVGL sounds like a good approach. I just looked at the way ZMK is doing things. I like the flexibility this gives in terms of changing the assignment with overlays. 

What do you think about the layering approach used in there? Maybe thats a bit much at the beginning - on the other hand it allows for really neat use cases with a very small keyboard matrix.

fabiobaltieri commented:
> Will do that then, seems to be worth the extra header imo.

Sorry I thought you were asking if you can include the main "lvgl.h" in a devicetree file, that would not work. But if you need codes yeah they defo go for a `dt-bindings/lvgl.h` file. What I'm saying it that you can only have defines there, no enums or other stuff, so if LVGL ships with a header with the data you need in an enum you'll have to redefine all the entries.

faxe1008 committed:
samples: subsys: display: lvgl: Add lvgl-keypad-input device

Adds a zephyr,lvgl-keypad-input compatible to the native_posix board
overlay and the required code to control an button matrix widget.

Signed-off-by: Fabian Blatz <fabianblatz@gmail.com>

fabiobaltieri reviewed (dismissed):
Cool, there's a compliance warning you prob want to look into.

faxe1008 commented on modules/lvgl/input/lvgl_keypad_input.c:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	if (i == cfg->num_codes) {
+		LOG_DBG("Ignored input event: %u", evt->code);
```
Will do :^)

faxe1008 commented:
Yeah I checked the compliance warning, however if I fix it, it will cause a `discards-qualifier` warning since the method signature looks like this `void lv_btnmatrix_set_map(lv_obj_t * obj, const char * map[])`

fabiobaltieri commented:
> Yeah I checked the compliance warning, however if I fix it, it will cause a `discards-qualifier` warning since the method signature looks like this `void lv_btnmatrix_set_map(lv_obj_t * obj, const char * map[])`

I see, that's annoying, there's two options:
- force push this, since this is checkpatch I think it only fires when you touch that line, so it may not be terribly painful to maintain
- declare it as checkpatch wants and cast it back on the call to make the compiler happy

what do you think?

fabiobaltieri reviewed (dismissed):
I have mixed feelings about suggesting that casting but I'll approve regardless. :-)

faxe1008 commented on samples/subsys/display/lvgl/src/main.c:
```diff
@@ -95,21 +100,36 @@ int main(void)
 	lv_indev_set_group(lvgl_input_get_indev(lvgl_encoder), arc_group);
 #endif /* CONFIG_LV_Z_ENCODER_INPUT */
 
+#ifdef CONFIG_LV_Z_KEYPAD_INPUT
```
True, thanks for reminding. Got a PR ready for you to review 🙃 : [https://github.com/zephyrproject-rtos/zephyr/pull/66256](https://github.com/zephyrproject-rtos/zephyr/pull/66256)
